### DISCUSSION TITLE
Backup of volumes in DMS

### DISCUSSION THREAD

**Discussion by ontheair81** [docker-mailserver/docker-mailserver#2554] [Q&A] [answered]
I have some questions about backing up volumes in DMS, as mentioned in the [FAQ](https://docker-mailserver.github.io/docker-mailserver/edge/faq/#volumes):

The shown example looks like this:

`docker run --rm -it \`
 ` -v "${PWD}/docker-data/dms/config/:/tmp/docker-mailserver/" \`
 ` -v "${PWD}/docker-data/dms-backups/:/backup/" \`
 ` --volumes-from mailserver \`
 ` alpine:latest \`
 ` tar czf "/backup/mail-$(date +%F).tar.gz" /var/mail /var/mail-state /var/logs/mail /tmp/docker-mailserver`

**1 - Typo**
I am quite sure that `/var/logs/mail` should be `/var/log/mail` instead (without "s").

**2 - Consistency**
These backups are done while DMS is running and I am bit scared that the backups might be inconsistent when operations are active while tar is working.
Is this a proven way to do backups and is this working reliable?

**3 - Understanding**
I don't get the need for the second line:
 ` -v "${PWD}/docker-data/dms/config/:/tmp/docker-mailserver/" \`
Maybe I am missing something, but for me it looks like this line is not needed.

For my crontabbed backup script I cut out this line, so it looks like this - and it seems to work fine:
`docker run --rm -it \`
 ` -v "/path/to/my/backup/folder/:/backup/" \`
`  --volumes-from mailserver \`
`  alpine:latest \`
`  tar czf "/backup/$(date '+%Y%m%d')-mailserver.tar.gz" /var/mail /var/mail-state /var/log/mail /tmp/docker-mailserver`

Am I missing something without this line (seems not to be the case)?

**4 - Restore**

The [FAQ shows how to restore backups when using bind mounts](https://docker-mailserver.github.io/docker-mailserver/edge/faq/#bind-mounts-default). But there is no information about how restoring is done when using volumes.

Of course there are [generic informations given by the official docker website](https://docs.docker.com/storage/volumes/#backup-restore-or-migrate-data-volumes), but in my opinion it would be helpful to include more specific informations in the FAQ of DMS.

I see myself as a beginner when it comes to docker and I guess I will be able to restore the backups in case this needs to be done - with some errors and cursing prior to success ;-)
It would be nice if there would be some short informations for guiding users and docker-none-experts ;-)

**Comment by casperklein** [comment, marked as answer]
Maybe the backup documentation needs some updates. Feel free to enhance it if you like 👍 

1. You are right.
2. To get a bullet proof consistent backup, I suggest to shutdown DMS for the backup. This ensures consistent data. When stopped, just make a copy of the directory `./docker-data/dms/`. All mails, logs, fail2ban database etc. is in there. Afterwards start DMS again. Use `docker-compose down/up` and not `docker-compose stop/start`.
3. `./docker-data/dms/config/` contains your configured email addresses, aliases etc.
4. All you need to do to restore your backup is to replaced `./docker-data/dms/` with your backup. You did no mention, which DMS version you are using. But for some time now, our [default config](https://github.com/docker-mailserver/docker-mailserver/blob/master/docker-compose.yml) only contains bind mounts. Volumes are not recommended anymore.

**Reply by ontheair81** [reply]
Thank you for these informations!

> Maybe the backup documentation needs some updates. Feel free to enhance it if you like

Still have to find out how to do this, but this is one of the points on my task list ;-)

> 4\. You did no mention, which DMS version you are using.

I upgraded to v10.5.0 as soon as it was available and running it since then.

> 4\. But for some time now, our [default config](https://github.com/docker-mailserver/docker-mailserver/blob/master/docker-compose.yml) only contains bind mounts. Volumes are not recommended anymore.

This is the most important information! Last year I installed DMS and intentionally changed the configuration to use volumes, because the [FAQ recommends to do so, as otherwise data may be lost](https://docker-mailserver.github.io/docker-mailserver/edge/faq/#where-are-emails-stored).

So I have to migrate from volumes to bind mounts.
- I guess this can be done by simply copying the contents of these volumes to their respective bind mounts while preserving the original timestamps and existing permissions of files and folders?
- All the content has to be chowned to root then, as docker is running as root? (EDIT: I see that different files/folders in the volumes have different permissions. Maybe these should be kept as is?).

**Reply by casperklein** [reply]
> Still have to find out how to do this, but this is one of the points on my task list ;-)

There is an edit button on each page of our documentation, it's pretty easy and straight forward to suggest changes. To format things, we use [Markdown](https://en.wikipedia.org/wiki/Markdown#Examples).

> I guess this can be done by simply copying the contents of these volumes to their respective bind mounts while preserving the original timestamps and existing permissions of files and folders?

Exactly. ~~You should find the volumes somewhere in `/var/lib/docker/volumes/`.~~ I guess you already figured that out.

>  I see that different files/folders in the volumes have different permissions. Maybe these should be kept as is

Yep. Just copy/move the files. Ownership/permissions should already be fine. Also DMS should take care of this on startup and will adjust permission if necessary.

**Reply by ontheair81** [reply]
This helps me a lot, thank you! I will migrate from volumes to bind mounts as soon as possible.
